#!/bin/bash 
#PBS -N mapgd-parallel-Bwa
#PBS -k o
#PBS -l nodes=1:ppn=16,walltime=96:00:00
#PBS -l vmem=100gb
#PBS -M ouqd@hotmail.com
#PBS -m abe
#PBS -j oe

set +x
echo PBS -N mapgd-debug
echo PBS -k o
echo PBS -l nodes=1:ppn=16,walltime=12:00:00
echo PBS -l vmem=100gb
echo PBS -M ouqd@hotmail.com
echo PBS -m abe
echo PBS -j oe
echo Updated on 05/28/2018
echo module rm gcc
echo module load gcc
echo module load gsl
echo module load samtools
module rm gcc
module load gcc
module load gsl  
module load samtools


export SampleID=PA2013
export DATA_DIR=/N/dc2/scratch/xw63/$SampleID/Bwa
export HeaderFile=$DATA_DIR/PA42.header

echo SampleID=PA2013
echo DATA_DIR=/N/dc2/scratch/xw63/$SampleID/Bwa
echo HeaderFile=$DATA_DIR/PA42.header

set -x
cd $DATA_DIR
set +x
echo ===============================================================
echo 0. Make proview files of nucleotide-read quartets -counts of A, C, G, and T, from each of the mpileup files of the clones of the population.
echo ===============================================================
echo This should be done before submitting this task by using the following commands:
echo  '              perl Make_pipeline.mapgd.pl'
echo  '              qsub ./mapgd_proview.pbs'
set +x

echo ===============================================================
echo 1. Combine all mapgd proview files into one.
echo ===============================================================
set -x
time java -cp ~/daphnia/DaphniaVariantCall CombineProview $DATA_DIR $SampleID
set +x
echo ===============================================================
echo 2. Exclude mtDNA data from the pro file.
echo ===============================================================
echo if mtDNA sequence is not included in the reference genome, skip this step:
echo if mtDNA sequence is included in the reference genome, execute this:
set -x
echo "time grep -v '^PA42_mt_genome' $SampleID.combined.pro.txt \> Nuc_$SampleID.combined.pro.txt"
echo "mv Nuc_$SampleID.combined.Pro.txt $SampleID.combined.pro.txt"
set +x

echo ===============================================================
echo 3. Run the allele command to estimate allele and genotype frequencies from the pro file.
echo ===============================================================
set -x
time mapgd allele -i $DATA_DIR/$SampleID.combined.pro.txt -o $DATA_DIR/$SampleID.combined.map -p $DATA_DIR/$SampleID.combined.clean
set +x

echo ===============================================================
echo 4. Run the filter command to filter the map file of ML estimates of the parameters.
echo ===============================================================
set -x
time mapgd filter -i $DATA_DIR/$SampleID.combined.map.map -p 20 -q 0.05 -Q 0.45 -c 800 -C 2400 -o $DATA_DIR/$SampleID.combined_filtered.map
set +x
echo -p: minimum value of the likelihood-ratio test statistic for polymorphism 
echo -q: minimum minor-allele frequency estimate
echo -Q: maximum minor-allele frequency estimate
echo -c: minimum population coverage
echo -C: maximum population coverage
echo ===============================================================
echo 5. Run the genotype command to generate a file of genotype likelihoods
echo ===============================================================
set -x
time mapgd genotype -p $DATA_DIR/$SampleID.combined.clean.pro -m $DATA_DIR/$SampleID.combined_filtered.map.map > $DATA_DIR/$SampleID.combined.genotype
set +x

echo ===============================================================
echo Remove the unnecessary header and footer
echo ===============================================================
set -x
time awk '{if ($3 != "MN_FREQ" && $3 >= 0.0 && $3 <= 1.0) print}' $DATA_DIR/$SampleID.combined.genotype > $DATA_DIR/$SampleID.combined_F.genotype
set +x
echo ===============================================================
echo Randomly pick a specified number - 200000 of SNPs from the file of genotype likelihoods 
===============================================================

set -x
time /N/dc2/projects/daphpops/Software/MAPGD-0.4.26/extras/sub_sample.py $DATA_DIR/$SampleID.combined_F.genotype -N 200000 > $DATA_DIR/$SampleID.combined_F_200K.genotype
set +x
echo ===============================================================
echo Extract the header from the file of genotype likelihoods
echo ===============================================================
set -x
time head -n -1 $DATA_DIR/$SampleID.combined.genotype | awk '{if ($3 == NULL || $1 ~ /^@/) print}' > $DATA_DIR/$SampleID.combined_header.genotype
set +x
echo ===============================================================
echo Extract the footer from the file of genotype likelihoods
echo ===============================================================

set -x
time tail -n 1 $DATA_DIR/$SampleID.combined.genotype > $DATA_DIR/$SampleID.combined_footer.genotype
set +x
echo ===============================================================
echo Add the header and footer to the sub-sample of the file of genotype likelihoods
echo ===============================================================

set -x
time cat $DATA_DIR/$SampleID.combined_header.genotype $DATA_DIR/$SampleID.combined_F_200K.genotype $DATA_DIR/$SampleID.combined_footer.genotype > $DATA_DIR/$SampleID.combined_F_200K_wh_wf.genotype
set +x
echo ===============================================================
echo 7. Run the relatedness command
echo ===============================================================

set -x
time mapgd relatedness -i $DATA_DIR/$SampleID.combined_F_200K_wh_wf.genotype -o $DATA_DIR/$SampleID.combined_F_200K_wh_wf_rel.out

set +x
echo ===============================================================
echo =============Task completed.===================
echo ===============================================================

